rust/src/model.rs

```diff
@@ -0,0 +1,569 @@
+use std::convert::TryFrom;
+
+use num::{bigint::BigInt, clamp, rational::Ratio, traits::identities::Zero};
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+use crate::{
+    mask::{config::MaskConfig, seed::MaskSeed, Integers, MaskedModel},
+    utils::generate_integer,
+    PetError,
+};
+
+/// Masking of models.
+pub trait MaskModels<N> {
+    /// Get a reference to the weights.
+    fn weights(&self) -> &Vec<N>;
+
+    /// Cast the weights as ratios. Must handle non-finite weights.
+    fn as_ratios(&self) -> Vec<Ratio<BigInt>>;
+
+    /// Mask the model wrt the mask configuration. Enforces bounds on the scalar and weights.
+    fn mask(&self, scalar: f64, config: &MaskConfig) -> (MaskSeed, MaskedModel) {
```

Comment:
Could the seed be an argument to the function? It would make testing this function easier if we can control the seed.

Reply:
i didn't put the seed as an argument to the method to avoid cases of "forgotten" seeding or seeding with non-random values. but i agree that it would help testing. what do you think would have more benefits?